Allow pre-release updates to specific dependencies

arixmkii opened the pull request:
Fixes #2634 

Added dedicated configuration, which alters pre-release version checks for specified dependencies.

Similar features has been mentioned there https://github.com/scala-steward-org/scala-steward/issues/1023#issuecomment-551942307

arixmkii commented:
I plan to update docs, but posting this early here to see if additional changes will be requested.

arixmkii commented:
I though it was, but adding more tests it didn't play nice with snapshots. Added support for snapshots and tests in a separate commit.

exoego reviewed (changes requested):
Thanks!
Code looks great to me.
Could you update document as well🙇
https://github.com/scala-steward-org/scala-steward/blob/main/docs/repo-specific-configuration.md

arixmkii commented:
Changes:
* Fixed compatibility with https://github.com/sbt/sbt-dynver versioning scheme for snapshots.
* Updated sample repo config with new feature documentation
* Added new config settings to all config settings test.

arixmkii commented:
I can squash commits and rebase if necessary. Kept commits separate to ease the review process.

armanbilge commented on docs/repo-specific-configuration.md:
```diff
@@ -52,6 +52,12 @@ updates.pin  = [ { groupId = "com.example", artifactId="foo", version = "1.1." }
 # Defaults to empty `[]` which mean Scala Steward will not ignore dependencies.
 updates.ignore = [ { groupId = "org.acme", artifactId="foo", version = "1.0" } ]
 
+# The dependencies which match the given patterns are allowed to be updated to pre-release from stable.
+#
+# Each pattern must have `groupId`, and may have `artifactId` and `version`.
+# Defaults to empty `[]` which mean Scala Steward will ignore all stable to pre-release update options.
+updates.allowPreRelease  = [ { groupId = "com.example", artifactId="foo" } ]
```
Question: if I add this, do I get Ms and RCs or do I get snapshots? What if only want Ms and RCs?

arixmkii replied:
For now both. Could you help me understand your requirements a bit better. Instead of boolean flag we would like to have pre-release modes like:

1. none - previous behavior
2. only Ms and RCs
3. only snapshots
4. all kind of pre-releases (2 + 3)

Is this correct? Or should it be 

1. none - previous behavior
2. only Ms and RCs
3. Ms, RCs and snapshots

It will end up having 2 lists in seconds anyway:
* updates.allowPreRelease
* updates.allowSnapshots

The question is "Should allowSnapshots include allowPreRelease?"

armanbilge replied:
Yes, I think `allowSnapshots` should imply `allowPreRelease`. Similarly, it should allow updates to ordinary stable releases! :) For example, if a pre-release or snapshot is succeeded by a stable release, it should update to that (at least until the next snapshot/pre-release is available).

exoego replied:
@arixmkii 
Could you add a test case for below?

> if a pre-release or snapshot is succeeded by a stable release, it should update to that

fthomas replied:
I agree with what @armanbilge said in https://github.com/scala-steward-org/scala-steward/pull/2635#discussion_r900295436. But I would prefer implementing `allowSnapshots` in a later PR so that this PR does not become too big. I think there is no harm if `allowPreRelease` implies `allowSnapshots` for now.

Btw, if we pluralize snapshots in `allowSnapshots` shouldn't we also pluralize pre-releases in `allowPreReleases`?

arixmkii replied:
Implemented requested changes:
* used plural version of preReleases, where appropriate
* added more tests and grouped them logically
* added filter to prohibit updates from regular pre-release to snapshot

Rebased to latest main and squashed commits.

exoego reviewed (approved):
LGTM 💯 
Sorry for late response.
